### DISCUSSION TITLE
GCP artifact registry access token

### DISCUSSION THREAD

**Discussion by devoteam-quentin-reynaud** [renovatebot/renovate#21975] [Request Help]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

35.71

### If you're self-hosting Renovate, select which platform you are using.

github.com

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

Successfully connect to my private GCP Artifact registry to update docker images with renovate.

### What you tried so far.

Hello, I try to use renovate to update my docker images in my private Artifact registry on GCP.
Renovate is running in GKE but it seems it can't get automatically credentials to connect to Artifact registry through the GKE metadata server.

I've installed gcloud sdk in the renovate docker image following the documentation [here](https://docs.renovatebot.com/docker/#using-short-lived-access-tokens). 
Here is my config in config.json
```
"hostRules": [
    {
      "matchHost": "europe-west1-docker.pkg.dev",
      "hostType": "docker",
      "token": "ARTIFACT_REGISTRY_ACCESS_TOKEN",
      "authType": "Bearer" 
    },
```

But I receive a 401 unauthorized error in logs.
However the token seems ok because when I try to curl the URL that is failing in logs with it, it works. 

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: GET https://europe-west1-docker.pkg.dev/v2/token?scope=repository%3AXXXXXXXXXX&service=europe-west1-docker.pkg.dev = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=401 retryCount=0, duration=25) (repository=XXXXXXXXXX)
DEBUG: Unauthorized docker lookup (repository=XXXXXXXXXXX)
       "registryHost": "https://europe-west1-docker.pkg.dev",
       "dockerRepository": "XXXXXXXXXXX/my-img"
DEBUG: Response code 401 (Unauthorized) (repository=XXXXXXXXXXX)
       "err": {
         "name": "HTTPError",
         "code": "ERR_NON_2XX_3XX_RESPONSE",
         "timings": {
           "start": 1683213862640,
           "socket": 1683213862640,
           "lookup": 1683213862641,
           "connect": 1683213862651,
           "secureConnect": 1683213862656,
           "upload": 1683213862656,
           "response": 1683213862664,
           "end": 1683213862665,
           "phases": {
             "wait": 0,
             "dns": 1,
             "tcp": 10,
             "tls": 5,
             "request": 0,
             "firstByte": 8,
             "download": 1,
             "total": 25
           }
         },
         "message": "Response code 401 (Unauthorized)",
         "stack": "HTTPError: Response code 401 (Unauthorized)\n    at Request.<anonymous> (/opt/buildpack/tools/renovate/35.71.0/node_modules/got/dist/source/as-promise/index.js:118:42)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)",
         "options": {
           "headers": {
             "user-agent": "RenovateBot/35.71.0 (https://github.com/renovatebot/renovate)",
             "accept": "application/json",
             "authorization": "***********",
             "accept-encoding": "gzip, deflate, br"
           },
           "url": "https://europe-west1-docker.pkg.dev/v2/token?scope=repository%3AsXXXXXXXXXXXXX%2Fmy-img%3Apull&service=europe-west1-docker.pkg.dev",
           "hostType": "docker",
           "username": "",
           "password": "",
           "method": "GET",
           "http2": false
         },
         "response": {
           "statusCode": 401,
           "statusMessage": "Unauthorized",
           "body": {
             "errors": [{"code": "UNAUTHORIZED", "message": "failed authentication"}]
           },
           "headers": {
             "content-type": "application/json",
             "x-content-type-options": "nosniff",
             "x-frame-options": "SAMEORIGIN",
             "x-xss-protection": "0",
             "date": "Thu, 04 May 2023 15:24:22 GMT",
             "content-length": "71",
             "connection": "close"
           },
           "httpVersion": "1.1",
           "retryCount": 0
         }
       }
DEBUG: Failed to get authHeaders for getTags lookup (repository=XXXXXXXXXXXXXXXXX)
DEBUG: Failed to look up docker package europe-west1-docker.pkg.dev
```

</details>

**Comment by rarkins** [comment]
Are you attempting to use a token from environment variables?

**Reply by devoteam-quentin-reynaud** [reply]
I'm replacing the ARTIFACT_REGISTRY_ACCESS_TOKEN key with a sed command, but I've also try to put my token directly in the config and same result :/

**Comment by viceice** [comment]
how do you use the token with the docker cli?

you probably need to configure it as password instead and use a specific username.

**Reply by devoteam-quentin-reynaud** [reply]
[Here](https://github.com/nielsbasjes/renovate/blob/4d3b3bd2495a1902b2a108e261a26f9f89212adb/lib/datasource/docker/common.ts#L84 ) it seems that when I set the token key, then it generate a request with the bearer token in the header. That's exactly what I'm doing with curl and it works but not in renovate. I know that there are specific username like` _json_key_base64` do you know if I need a special one to use my GCP access token ?

**Reply by viceice** [reply]
please answer my question.

how do you use that token with the docker cli to login?

**Reply by devoteam-quentin-reynaud** [reply]
Like this `gcloud auth print-access-token | docker login -u oauth2accesstoken \
    --password-stdin https://europe-west1-docker.pkg.dev`
So according what you said I tried to set the username to oauth2accesstoken and add the token as password but same result.

**Reply by viceice** [reply]
it should work with:
```json
"hostRules": [
    {
      "matchHost": "europe-west1-docker.pkg.dev",
      "hostType": "docker",
      "username": "oauth2accesstoken",
      "password": "<replace-with-access-token>"
    },
```
